### DISCUSSION TITLE
Does it make sense to use ThreadPoolExecutor in fastAPI

### DISCUSSION THREAD

**Discussion by hjilke** [fastapi/fastapi#6666] [Questions] [answered]
Hi All,

I have a super simple app that has only one endpoint. This endpoint loads data from a database and this is parallelised
using a ThreadPoolExecutor. For example:

```
@app.get('/load_from_db')
def load_from_db():
    ....
    with concurrent.futures.ThreadPoolExecutor() as executor:
        for dataset in datasets:
            executor.submit(dataset.load_from_database)
    ....
```
Now I changed from Flask to FastAPI. I have declared my function as `def load_from_db` such that it is executed in a different thread-pool and does not block the main thread.

**OUT-DATED**
As a result my service now is 10x slower than using Flasks? I tried to set max_workers=5 but did no really help. What is the reason?

**EDIT:**
I created again some thorough test cases and it turns out fastAPI is not slower for me. Let's change the question to:

**Is it safe & does it make sense to use ThreadPoolExecutor in fastAPI?**

**Comment by hellocoldworld** [comment]
Wild guess, but in FastAPI you're already running the endpoint in a different thread, so the main thread is not blocked. So you can get rid of the `concurrent.futures.ThreadPoolExecutor`.

**Comment by nzig** [comment, marked as answer]
@hellocoldworld is correct. 
I will add that in some cases you do actually need a thread pool, if you want to mix async with blocking operations. In that case you can use Starlette's `run_in_threadpool`:
```python
from starlette.concurrency import run_in_threadpool

@app.get('/')
async def():
   await do_something_async()
   await run_in_threadpool(do_something_blocking)
```

This is actually what FastAPI uses internally if you define your handler as `def` (and not `async def`).

**Comment by hjilke** [comment]
Do you think there is a performance disadvantage of using a ThreadPoolExecuter in a regular `def` function in FastAPI?

**Comment by nzig** [comment]
Yes, it's slightly slower. If you use `ThreadPoolExecuter` in a `def` function in FastAPI, what happens is:

1. Thread 1 runs FastAPI. Since FastAPI is an async application and your `def` function might block, it calls it with `run_in_threadpool`, which runs it in a thread 2.
2. Thread 2 runs your function. Since you created a `ThreadPoolExecuter`, this creates thread 3.
3. Thread 3 runs whatever you did with the `ThreadPoolExecuter`.

This means for the handler to complete you need 4 thread switches (1->2->3->2->1).

If you use an `async def` there are 0 thread switches, and if you use `async def` with `run_in_threadpool` there are 2. Since each thread switch adds overhead, using `ThreadPoolExecuter` inside a `def` function will probably be slower.

**Comment by hjilke** [comment]
Thanks for all the answers!

**Comment by shawnwall** [comment]
Old conversation... but I don't quite understand how the answer here solves the problem. Given @hjilke was trying to execute multiple db loads at once, and was not using an `async` endpoint, how could he just remove the usage of the `ThreadPoolExecutor` and have have the multiple reads run concurrently? Each db load call would be synchronous.

**Comment by HuskyDanny** [comment]
Thanks @nzig and @hjilke, I have been struggling about this, This is the very good discussion point on the internal of fastapi about making the sync async.

I get for the path function - fastapi will run threadpool to execute the function in another thread and make it async.
But if I have an async path function, and in there I accidentally introduces some blocking function? Will it impact the event loop making it slow than it is expected? If fastapi can still handle, how fastapi identity a piece of code can be ascyned from sync?

**Comment by YannickGibson** [comment]
I am getting `greenlet.error: cannot switch to a different thread` using playwright

**Reply by hellocoldworld** [reply]
is this something related to the original discussion? If yes, please tell us how. If not, you should open a new discussion, or, if not related to fastapi at all, try a more general purpose forum like StackOverflow.
